### ISSUE TITLE
OLD  class-customization concept exercise

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2350]
This issue describes how to implement the `class-customization` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/use of `class-customization`  in Python. 

## Learning objectives

- more fully understand the uses (_and possible abuses_) of `classes` in Python
- understanding of `class members`
- understand some of the options that can be used to adapt a `class` for purposes beyond simple `object` creation, including:
  - using `@staticmethod` to provide namespaced functionality without  `object` instantiation.
  - using  `@classmethod` to provide functionality that modifies `class`  attributes shared across  `class instances`
  - using `__repr__()` and `__str__()` to provide human-readable and understandable class descriptions.

- using `property()` and `@property`  to make getters and setters for the methods of a `class`
  - using `property()` and `@property` to "compose" or "calculate" and return `instance attributes`.
  - using `property()` and `@property` to "protect" an attribute from mutation.
- using [`functools.cached_property()`](https://docs.python.org/3.9/library/functools.html#functools.cached_property) to calculate a property via a _method_, and the cache it for future calls to the property.

- Pythons version of "non-public" methods and "private variables" (AKA `Name mangling`)

## Out of scope

- `class-inheritance`, `multiple-inheritance`, `__super()__`, class `mixins`
- `class-composition`  -- (_beyond the composition needed for the decorators in this exercise_)
- `dataclasses`
- `decorators` outside of `@property`, `@staticmethod`, and `@classmethod` (_the student should already be exposed to the function version of these in a prerequisite_)
- `generators`
- `coroutines`
- `descriptors` (_these will get their own exercise_)
- using a `class` as a decorator
- performance considerations

  

## Concepts

- `classes`
- `class attributes`
- `class members`
- `class methods`, `@classmethod`
- `static methods`, `@staticmethod`
- `decorators`
- `getters` & `setters` (through the use of `property()` and `@property`)
- `instance attributes`
- `instance methods`
- `non-public methods`, `private variables`, and `name-mangling`

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `booleans`
- `classes`
- `comparisons`
- `decorators`
- `dicts`
- `functions`
- `higher-order-functions`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

  

## Resources to refer to

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [private variables (Python tutorial)](https://docs.python.org/3/tutorial/classes.html#tut-private)
- [Python Data Model - Python Docs](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy)
- [class property() - (Python Docs)](https://docs.python.org/3/library/functions.html#property)
- [Real Python:  Python's Instance, Class, and Static Methods Demystified](https://realpython.com/instance-class-and-static-methods-demystified/)
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  * You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  *  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  *  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson) & [concept exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md)

* ### Concept `about.md`

  **Concept file/issue**: There is currently no issue or files for the concept.  **They are TBD.**

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  * This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.
 

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** : `79af6c54-c826-4971-833f-2039dd6982bc`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
  Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.
- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
Currently being worked on in PR#2960.

**Comment by BethanyG** [comment]
closing for rewrite
